src/protocols/user_directed_commissioning/UserDirectedCommissioning.h

```diff
@@ -0,0 +1,192 @@
+ *      This file defines objects for a CHIP Echo unsolicitied
+ *      initaitor (client) and responder (server).
+ *
+ */
+
+#pragma once
+
+#include "UDCClients.h"
+#include <core/CHIPCore.h>
+#include <mdns/Resolver.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <messaging/Flags.h>
+#include <protocols/Protocols.h>
+#include <support/CodeUtils.h>
+#include <support/DLLUtil.h>
+#include <support/logging/CHIPLogging.h>
+#include <transport/TransportMgr.h>
+
+namespace chip {
+namespace Protocols {
+namespace UserDirectedCommissioning {
+
+/**
+ * User Directed Commissioning Protocol Message Types
+ */
+enum class MsgType : uint8_t
+{
+    IdentificationDeclaration = 0x00,
+};
+
+class DLL_EXPORT InstanceNameResolver
+{
+public:
+    /**
+     * @brief
+     *   Called when a UDC message is received specifying the given instanceName
+     * This method indicates that UDC Server needs the Commissionable Node corresponding to
+     * the given instance name to be found. UDC Server will wait for OnCommissionableNodeFound.
+     *
+     * @param instanceName DNS-SD instance name for the client requesting commissioning
+     *
+     */
+    virtual void FindCommissionableNode(char * instanceName) {}
+
+    virtual ~InstanceNameResolver() {}
+};
+
+class DLL_EXPORT UserConfirmationProvider
+{
+public:
+    /**
+     * @brief
+     *   Called when a UDC message has been received and corresponding nodeData has been found.
+     * It is expected that the implementer will prompt the user to confirm their intention to
+     * commission the given node, and obtain the setup code to allow commissioning to proceed,
+     * and then invoke commissioning on the given Node (using CHIP Device Controller, for example)
+     *
+     * @param nodeData DNS-SD node information for the client requesting commissioning
+     *
+     */
+    virtual void OnUserDirectedCommissioningRequest(const Mdns::DiscoveredNodeData & nodeData) {}
+
+    virtual ~UserConfirmationProvider() {}
+};
+
+class DLL_EXPORT UserDirectedCommissioningClient
+{
+public:
+    /**
+     * Send a User Directed Commissioning message to a CHIP node.
+     *
+     * @param transportMgr  A transport to use for sending the message.
+     * @param payload       A PacketBufferHandle with the payload.
+     * @param commissioner  Address of destination.
+     * @param port          Port of destination.
+     *
+     * @return CHIP_ERROR_NO_MEMORY if allocation fails.
+     *         Other CHIP_ERROR codes as returned by the lower layers.
+     *
+     */
+
+    CHIP_ERROR SendUDCMessage(TransportMgrBase * transportMgr, System::PacketBufferHandle && payload,
+                              chip::Inet::IPAddress commissioner, uint16_t port);
+};
+
+class DLL_EXPORT UserDirectedCommissioningServer : public TransportMgrDelegate
+{
+public:
+    static UserDirectedCommissioningServer & GetInstance()
+    {
+        static UserDirectedCommissioningServer instance;
```

Comment:
How does this work with CHIP stack shutdown and restart?

Reply:
Removed the singleton pattern